### PR TITLE
Feat/v3 hanging protocol MxN

### PR DESCRIPTION
**Pull request by wayfarer3130**
There were a number of defects in actually creating usable hanging
protocols.  This feature fixes a couple of them, and adds two hanging
protocols in the default theme.json5 file.

@sedghi - I think you are the person knowing something about hanging protocols.  I added a few extra options to the hanging protocols, including the ability to have them site-configured on a per-url basis (eg to allow different users to get different protocols).  They are also somewhat easier to configure than they used to be.  Should I have used stages to manage the hanging protocol setup for this?  Where should I document the changes?

### CODE REVIEW

File: platform/core/src/services/HangingProtocolService/HPMatcher.js [hunk 1]
```diff
@@ -36,7 +36,13 @@ const match = (metadataInstance, rules, customAttributeRetrievalCallbacks) => {
 
     // Create a single attribute object to be validated, since metadataInstance is an
     // instance of Metadata (StudyMetadata, SeriesMetadata or InstanceMetadata)
-    const attributeValue = metadataInstance[attribute];
+    let attributeValue = metadataInstance[attribute];
```
Comment (wayfarer3130) [inline platform/core/src/services/HangingProtocolService/HPMatcher.js:39]: The study level information is very sparse, because it doesn't actually start with a study query, so generate some of the study level information automatically.

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js [hunk 2]
```diff
@@ -35,11 +36,21 @@ class HangingProtocolService {
...
+    this.addProtocols(point.protocols);
+  }
+
   reset() {
     this.studies = [];
     this.protocols = [];
     this.hpAlreadyApplied = [];
     this.matchDetails = [];
+
+    if (!this._listenThemeProtocols) {
```
Comment (wayfarer3130) [inline platform/core/src/services/HangingProtocolService/HangingProtocolService.js:50]: Load any protocols specified by the theme properties, which allows setting up custom sets of hanging protocols for specific uses or users - that is, as a user configurable item rather than hard coded in the extensions modules.

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js [hunk 3]
```diff
@@ -54,6 +65,21 @@ class HangingProtocolService {
   addProtocols(protocols) {
     protocols.forEach(protocol => {
       if (this.protocols.indexOf(protocol) === -1) {
+        protocol.id = protocol.id || protocol.name;
```
Comment (wayfarer3130) [inline platform/core/src/services/HangingProtocolService/HangingProtocolService.js:68]: This is just a bunch of code to allow easier configuration of the hanging protocol by defining fewer bits of it.

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js [hunk 4]
```diff
@@ -314,7 +342,12 @@ class HangingProtocolService {
 
       highestStudyMatchingScore = studyMatchDetails.score;
 
+      // TODO - use display sets instead of series, as that will use the
+      // correct re-grouped information.
       study.series.forEach(aSeries => {
+        if (existingMatches && existingMatches.indexOf(aSeries) != -1) {
```
Comment (wayfarer3130) [inline platform/core/src/services/HangingProtocolService/HangingProtocolService.js:348]: Allows matching consecutive series, without re-displaying the same one multiple times.  Need to figure out a good way to allow re-matching an existing instance.  I was thinking of adding viewport rules that will find a previous series again.

**Comment by sedghi** [comment]
I will take a look into this, maybe let's wait until my review and then you add to documentations?

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js [hunk 5]
```diff
@@ -314,7 +342,12 @@ class HangingProtocolService {
 
       highestStudyMatchingScore = studyMatchDetails.score;
 
+      // TODO - use display sets instead of series, as that will use the
```
Comment (wayfarer3130) [inline platform/core/src/services/HangingProtocolService/HangingProtocolService.js:345]: Changing the matching here would allow interaction with the display sets, which means that proper regrouping rules could be defined, and then directly used by the hanging protocols.  That ultimately makes for much easier to define HPs because you don't have to specify both the regrouping rules as well as the hanging rules.  HOwever, compatibility with the existing DICOM HPs should be maintained, so need to think about how to do this properly.

File: platform/viewer/src/appInit.js [hunk 6]
```diff
@@ -17,6 +17,7 @@ import {
   errorHandler
   // utils,
 } from '@ohif/core';
+import ConfigPoint from 'config-point';
```
Comment (sedghi) [inline platform/viewer/src/appInit.js:20]: Taking a quick look, I saw you are using config-point, is this intended or mistakenly put here? since we haven't reviewed the config-point additions before

File: platform/viewer/public/theme/theme.json5 [hunk 7]
```diff
@@ -0,0 +1,80 @@
+{
+  ThemeProtocols: {
```
Comment (wayfarer3130) [inline platform/viewer/public/theme/theme.json5:2]: It is possible to define additional hanging protocols for different uses at a site by defining more themes and using theme=MoreHangingProtocols

File: platform/viewer/public/theme/theme.json5 [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    protocols: [
+      {
+        id: '1x2',
```
Comment (wayfarer3130) [inline platform/viewer/public/theme/theme.json5:5]: Note how this is somewhat easier to define - it has a bunch of stuff defaulted or not required.

File: platform/viewer/src/appInit.js [hunk 9]
```diff
@@ -49,6 +50,14 @@ function appInit(appConfigOrFunc, defaultExtensions) {
     appConfig,
   });
 
+  if (window) {
```
Comment (wayfarer3130) [inline platform/viewer/src/appInit.js:53]: This stuff loads the theme configuration values at startup, defaulting to the theme.json5 is not otherwise specified.

File: platform/viewer/src/appInit.js (hunk 6 above)
Comment (wayfarer3130) [reply platform/viewer/src/appInit.js:20]: It is intended to be used - it allows a site to reasonably define application specific extensions.

Comment (sedghi) [reply platform/viewer/src/appInit.js:20]: I don't think this the best approach. This PR currently adds the `config-points` dependencies (which is an external 3rd party package dependency on your Github) to a lot of core code. First of all, I think if we are going ahead with `config-point` which we were going to discuss, and it should have its own PR; for transparency and maintainability, it should be an OHIF package (something like platform/config-point). The same is true for the static-wado stuff which we need to move to @cornerstone (which I think we have talked about it previously).

Hanging protocols are somehow flexible right now, and they can be set for specific user (for instance fetching from a server user-specific HP's), and they can be added in the ModeEnter to the HangingProtocolService (`HangingProtocolService.setHangingProtocol()`), so extensions are not the only place providing the protocols for matching, Mode's can manipulate the HP (and that is the intension of Modes, PT/CT mode: 3x3 layout, Prostate MRI 2x2 layout etc.). What is config-point doing for enhancement of HP stuff?

CC @swederik

Comment (wayfarer3130) [reply platform/viewer/src/appInit.js:20]: I'm fine with config-point having it's own PR, but it can't be an OHIF package since it isn't OHIF specific, but is intended to be used in other contexts (in fact I'm already using it elsewhere).  
modes are defined in extensions, so whether the additional HPs are in the mode or are in the base extension, they are still defined elsewhere, and that means modifying code in order to provide them.  The design for ConfigPoint is that it is an easily included configuration system that allows having several variations on the configuration, without needing to include services that are not commonly available, such as the DICOM Hanging Protocol service - which exists in a couple of places but is very cumbersome to use and will mostly require sites adding custom code in order to support it.  config-point just needs static files added, so it meets the same sort of requirements for deployment that you are going to have for OHIF anyways.  The files can be gzipped on disk and served already compressed, they can be edited in any plain text editor and redeployed without rebuilding any of the software.  That is being used already for specific configurations deployed to different teams within a healthcare setting (FU Berlin specifically, different deployments for each department, and for each species being treated).
In terms of use with modes, I actually agree there - the modes should have names/settings for specific use cases, but should also be site definable - not require code level customizations to setup.  If a site wants the PT/CT mode to auto-open a fusion viewport, and then two other viewports, well, that becomes configurable at the site, as long as the underlying code has it supported.

Comment (wayfarer3130) [reply platform/viewer/src/appInit.js:20]: I'm fine with config-point having it's own PR, but it can't be an OHIF package since it isn't OHIF specific, but is intended to be used in other contexts (in fact I'm already using it elsewhere).
modes are defined in extensions, so whether the additional HPs are in the mode or are in the base extension, they are still defined elsewhere, and that means modifying code in order to provide them. The design for ConfigPoint is that it is an easily included configuration system that allows having several variations on the configuration, without needing to include services that are not commonly available, such as the DICOM Hanging Protocol service - which exists in a couple of places but is very cumbersome to use and will mostly require sites adding custom code in order to support it. config-point just needs static files added, so it meets the same sort of requirements for deployment that you are going to have for OHIF anyways. The files can be gzipped on disk and served already compressed, they can be edited in any plain text editor and redeployed without rebuilding any of the software. That is being used already for specific configurations deployed to different teams within a healthcare setting (different deployments for each department, and for each species being treated).
In terms of use with modes, I actually agree there - the modes should have names/settings for specific use cases, but should also be site definable - not require code level customizations to setup. If a site wants the PT/CT mode to auto-open a fusion viewport, and then two other viewports, well, that becomes configurable at the site, as long as the underlying code has it supported.

File: extensions/cornerstone/src/commandsModule.js [hunk 10]
```diff
@@ -291,32 +291,6 @@ const commandsModule = ({ servicesManager, commandsManager }) => {
         });
       }
     },
-    updateTableWithNewMeasurementData({
```
Comment (sedghi) [inline extensions/cornerstone/src/commandsModule.js:294]: why is this removed?

File: platform/core/src/services/HangingProtocolService/HPMatcher.js [hunk 11]
```diff
@@ -7,7 +7,7 @@ import validate from './lib/validator';
  * @param  {Array} rules Array of MatchingRules instances (StudyMatchingRule|SeriesMatchingRule|ImageMatchingRule) for the match
  * @return {Object}      Matching Object with score and details (which rule passed or failed)
  */
-const match = (metadataInstance, rules, customAttributeRetrievalCallbacks) => {
+const match = (metadataInstance, rules, customAttributeRetrievalCallbacks, existing) => {
```
Comment (sedghi) [inline platform/core/src/services/HangingProtocolService/HPMatcher.js:10]: can you update the documentation of this function too?

File: platform/core/src/services/HangingProtocolService/HPMatcher.js [hunk 12]
```diff
@@ -36,7 +36,13 @@ const match = (metadataInstance, rules, customAttributeRetrievalCallbacks) => {
 
     // Create a single attribute object to be validated, since metadataInstance is an
     // instance of Metadata (StudyMetadata, SeriesMetadata or InstanceMetadata)
-    const attributeValue = metadataInstance[attribute];
+    let attributeValue = metadataInstance[attribute];
+    if (attributeValue === undefined) {
+      if (attribute == 'NumberOfStudyRelatedSeries') {
```
Comment (sedghi) [inline platform/core/src/services/HangingProtocolService/HPMatcher.js:41]: === instead of ==

File: platform/docs/docs/configuration/configuration.md [hunk 13]
```diff
@@ -59,6 +73,38 @@ window.config = {
...
+parameters, for example:
+```
+https://ohif.hospital.org/?theme=mgHP&theme=euroKeyboard
+```
+to load two hypothetical theme settings files mgHP and euroKeyboard to add
+mammographic hanging protocols and European keyboard settings.
+
+A site can add such settings by creating custom files in:
+```
+.../build/theme/mgHP.json5
```
Comment (sedghi) [inline platform/docs/docs/configuration/configuration.md:90]: can we be more specific about where these themes should be located? where is `build` located? how to experiment with it in the `development`

File: platform/docs/docs/configuration/configuration.md [hunk 14]
```diff
@@ -59,6 +73,38 @@ window.config = {
...
+.../build/theme/euroKeyboard.json5
+```
+A site might build such different themes to support various user preferences
+or site differences between users, such as themes to support specific clinics
+or differences in user groups such as left on right mammography viewing versus
+right on left mammography viewing.
+
+The decision to use the JSON5 perser for this was primarily aimed at allowing
```
Comment (sedghi) [inline platform/docs/docs/configuration/configuration.md:98]: parser

File: platform/docs/docs/configuration/theme-configuration.md [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+
+## Config-Point
+The configuration schema is based on the [config-point](https://github.com/OHIF/config-point)
+library.  This library is design to allow developers to creation configuration
+points for their code by declaring static values, which can be modified externally.
+See [config-point-service](../platform/services/config-point-service.md) for
+internal details and development documentation.
+
+## Theme Files
+Theme files are simple json5 definition files, placed into the deployment
```
Comment (sedghi) [inline platform/docs/docs/configuration/theme-configuration.md:13]: same here, the location of the theme files are not clear to me

File: platform/docs/docs/configuration/theme-configuration.md [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+  },
+}
+```
+as an example that modifies the hanging protocols, both by adding and updating
+elements.  Note how this is a deep modification to a value.  The intent is
+to allow modifying configuration values which are heavily nested and/or list
+based.
+
+## Working with Objects
+The object tree is matched by the simple attribute name.  The attribute tree
```
Comment (sedghi) [inline platform/docs/docs/configuration/theme-configuration.md:53]: I understadn what this section is, but I needed to read twice to figure out that we are explaining how config-point works here (first I thought this is continuation of theme and hp stuff)

File: platform/docs/docs/configuration/theme-configuration.md [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+
+### Immediate operations
+Immediate operations perform the action immediately, and can thus be
+extended further.  They support the additional keys:
+* position to modify something at a given position
+
+The immediate operations are:
+* replace, insert, delete
+
+For example, supposing there was a configuraiton point ModalityList, then
```
Comment (sedghi) [inline platform/docs/docs/configuration/theme-configuration.md:125]: typo in configurations

File: platform/docs/docs/configuration/theme-configuration.md [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+// Delete an item with value or id 'MR'
+ModalityList: {
+  'MR': {configOperation: 'delete'},
+
+// Insert before the item with id 'MG'
+ModalityList: {
+  'MG': {configOperation: 'insert', value: 'MGTomo'},
+```
+
+### Getter Opreations
```
Comment (sedghi) [inline platform/docs/docs/configuration/theme-configuration.md:145]: typo in operations

File: platform/docs/docs/configuration/theme-configuration.md [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+```
+The MG protocol doesn't initially exist, so this would add a new hanging
+protocol, which would be defined in the normal hanging protocol definition.
+
+## Query List
+One of the suggested areas for customization is the columns in the query table.
+TODO
+
+## Demographics Overlay
+Another recommended change is to configure the demographcis overlay using
```
Comment (sedghi) [inline platform/docs/docs/configuration/theme-configuration.md:233]: typo demographics

File: platform/viewer/public/theme/theme.json5 [hunk 20]
```diff
@@ -0,0 +1,2 @@
+{
```
Comment (sedghi) [inline platform/viewer/public/theme/theme.json5:1]: is this intentionally empty? can we have an actual hp here so that the demo can showcase the `confi-point` stuff?

File: platform/core/src/services/HangingProtocolService/themeProtocolProvider.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
+import ConfigPoint from 'config-point';
+
+const { ThemeProtocols } = ConfigPoint.register({
```
Comment (sedghi) [inline platform/core/src/services/HangingProtocolService/themeProtocolProvider.js:3]: why are we specifying hangingProtocols inside the `core`?

File: platform/viewer/src/appInit.js [hunk 22]
```diff
@@ -49,6 +50,14 @@ function appInit(appConfigOrFunc, defaultExtensions) {
...
+    // Make the config point accessible in the window
+    window.ConfigPoint = ConfigPoint;
```
Comment (sedghi) [inline platform/viewer/src/appInit.js:55]: is this window assignment intentional?

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js [hunk 23]
```diff
@@ -54,6 +65,21 @@ class HangingProtocolService {
   addProtocols(protocols) {
     protocols.forEach(protocol => {
       if (this.protocols.indexOf(protocol) === -1) {
+        protocol.id = protocol.id || protocol.name;
+        protocol.name = protocol.name || protocol.id;
+        const { stages } = protocol;
+
+        stages.forEach(stage => {
```
Comment (sedghi) [inline platform/core/src/services/HangingProtocolService/HangingProtocolService.js:72]: I know what you are doing here, but maybe put some comment of why for people reading this

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js [hunk 24]
```diff
@@ -227,6 +254,7 @@ class HangingProtocolService {
       // Viewport Settings
       //
       // protocol defined callback
+      viewport.viewportSettings = viewport.viewportSettings || { filter: () => [] };
```
Comment (sedghi) [inline platform/core/src/services/HangingProtocolService/HangingProtocolService.js:257]: default viewportSettings can be just an array (no need to define `filter` as Arra.filter exist)

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js (hunk 5 above)
Comment (sedghi) [reply platform/core/src/services/HangingProtocolService/HangingProtocolService.js:345]: what do you mean by "interaction with display sets"?

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js (hunk 4 above)
Comment (sedghi) [reply platform/core/src/services/HangingProtocolService/HangingProtocolService.js:348]: what if it is fusion? pt-axial, ct-axial, fusion-axial (both pt-axial, ct-axial)

Comment (sedghi) [reply platform/core/src/services/HangingProtocolService/HangingProtocolService.js:348]: we are actually moving to `displaySets` in future (because of fusion stuff), so I would suggest to stick with the series matching by rules only

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js [hunk 25]
```diff
@@ -357,6 +390,7 @@ class HangingProtocolService {
           SeriesInstanceUID: aSeries.SeriesInstanceUID,
           matchingScore: totalMatchScore,
           matchDetails: matchDetails,
+          displaySet: aSeries,
```
Comment (sedghi) [inline platform/core/src/services/HangingProtocolService/HangingProtocolService.js:393]: I'm not sure if `displaySet` makes sense here, since this matching function should only care about metadata matching and sorting them. Why do we need `displaySet` here?

File: platform/core/src/services/HangingProtocolService/HangingProtocolService.js [hunk 26]
```diff
@@ -387,7 +421,7 @@ class HangingProtocolService {
     );
 
     const bestMatch = matchingScores[0];
-
+    if (bestMatch?.displaySet) existingMatches.push(bestMatch.displaySet);
```
Comment (sedghi) [inline platform/core/src/services/HangingProtocolService/HangingProtocolService.js:424]: If for any reason we end up using the existingMatches (which I explained why we shouldn't above), this push to array should happen outside of `match` function

File: platform/viewer/src/appInit.js (hunk 6 above)
Comment (sedghi) [reply platform/viewer/src/appInit.js:20]: we can resolve this i guess after the new commits

**Comment by wayfarer3130** [comment]
This will be revisited when the cornerstone 3d changes are integrated.
